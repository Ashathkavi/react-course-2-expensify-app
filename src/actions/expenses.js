import uuid from 'uuid'
import database from '../firebase/firebase'


//ADD_EXPENSE
export const addExpense = (expense) => ({
    type:'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData ={}) => {
    return (dispatch)=>{
        const { description = '', note= '', amount=0, createdAt=0 } = expenseData
        const expense = {description , note, amount, createdAt}
        return database.ref('expenses').push(expense)
            .then((ref)=>{
                dispatch(addExpense({
                    id:ref.key,
                    ...expense
                }))
            })
    }
}
 

//REMOVE_EXPENSES
export const removeExpense = ({id} = {}) => ({
    type:'REMOVE_EXPENSE',
    id
})


//ASYNC REMOVE_EXPENSES

export const startRemoveExpenses = ({id}={}) => {
    return (dispatch)=>{
        //const { description = '', note= '', amount=0, createdAt=0 } = expenseData
        //const expense = {description , note, amount, createdAt}
        return database.ref(`expenses/${id}`).remove()
            .then(()=>{
                dispatch(removeExpense({id}))
            }).catch((error)=>{
                console.log("Remove failed: " + error.message)
            })                     
    }
}



//EDIT_EXPENSES
export const editExpense = (id, updates) => ({
    type:'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpenses = (id, updates) => {
    return (dispatch)=>{
        return database.ref(`expenses/${id}`).update({
            ...updates
        }).then(()=>{
                dispatch(editExpense(id, updates))
        }).catch((error)=>{
            console.log("Edit failed: " + error.message)
        })                     
    }
}



//SET_EXPENSES
export const setExpenses = (expenses) => ({
    type:'SET_EXPENSES',
    expenses
})


//ASYNC SET_EXPENSES

export const startSetExpenses = () => {
    return (dispatch)=>{
        //const { description = '', note= '', amount=0, createdAt=0 } = expenseData
        //const expense = {description , note, amount, createdAt}
        return database.ref('expenses')
            .once('value').then((snapshot)=>{
                const expenses = []
                snapshot.forEach((childSnapshot)=>{
                    expenses.push({
                        id:childSnapshot.key,
                        ...childSnapshot.val()
                    })
                })
                dispatch(setExpenses(expenses))
            }, (e)=>{
                console.log('Error with data fetching: ', e)
            })            
    }
}