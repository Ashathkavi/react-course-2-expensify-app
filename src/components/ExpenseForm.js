import React from 'react'
import moment from 'moment'
import DatePicker from 'react-datepicker'


const now = moment()
//console.log(now.format('MMM do, YYYY'))

class ExpenseForm extends React.Component{
    constructor(props){
        super(props)
        this.state= {
            description:props.expense ? props.expense.description : '', 
            note:props.expense ? props.expense.note : '', 
            amount:props.expense ? (props.expense.amount / 100).toString() : '', 
            createdAt:props.expense ?  moment(props.expense.createdAt).toDate() : moment().toDate(),
            error:''
        }
        
    }
    
    
    onDescriptionChange = (e) => { 
        e.persist()
        this.setState(()=>({description:e.target.value}))
    }
    onNoteChange = (e) => {
        e.persist()
        this.setState(()=>({note:e.target.value}))
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/gm)){
            this.setState(()=>({amount}))
        }        
    }
    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({createdAt}))
        }
    }
    onFocusChange = ({focused}) => {
        this.setState(()=>({calenderFocused:focused}))
    }
    onSubmit = (e) => {
        e.preventDefault()
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error:'Please provide description and amount.'}))
        }else{
            this.setState(()=>({error:''}))
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount, 10) * 100,
                createdAt:this.state.createdAt.valueOf(),
                note:this.state.note
            })
        }
    }
    render(){
        //console.log(this.state.description)
        return(
            <div>
            {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit = {this.onSubmit}>
                    <input 
                        typr="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}                        
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}                        

                    />
                    <DatePicker
                        selected = {this.state.createdAt}
                        onChange = {this.onDateChange}
                        closeOnScroll={e => e.target === document}
                    />
                    <textarea 
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>

                </form>
            </div>
        )
    }
}

export default ExpenseForm