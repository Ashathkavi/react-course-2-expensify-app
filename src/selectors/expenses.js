
import moment from 'moment'


//get visible expenses
const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    //console.log(startDate, "startDate")
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMoment = moment(startDate)
        const endDateMoment = moment(endDate)
        //console.log(startDateMoment,'startDateMoment')
        // console.log(createdAtMoment,'createdAtMoment')
        // console.log(endDateMoment,'endDateMoment')

        const startDateMatch = startDate ? startDateMoment.isSameOrBefore(createdAtMoment, 'day'):true
        //const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = endDate ? endDateMoment.isSameOrAfter(createdAtMoment, 'day'):true
        //const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endDateMatch && textMatch

    }).sort((a,b)=>{
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1
        }else if(sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1
        }
    })
}

export default getVisibleExpenses