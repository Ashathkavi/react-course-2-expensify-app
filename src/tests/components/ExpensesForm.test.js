import React from 'react'
import {shallow} from 'enzyme'
import ExpenseForm from '../../components/ExpenseForm'
import expenses from '../fixtures/expenses'
import DatePicker from 'react-datepicker'


test('should render ExpenseForm correctly', ()=>{
    const wrapper = shallow(<ExpenseForm />)
    expect(wrapper).toMatchSnapshot()
})


test('should render ExpenseForm with expense data', ()=>{
    const wrapper = shallow(<ExpenseForm expense={expenses[0]}/>)
    expect(wrapper).toMatchSnapshot()
})


test('should render error for invalid form submission', ()=>{
    const wrapper = shallow(<ExpenseForm/>)
    expect(wrapper).toMatchSnapshot()
    wrapper.find('form').simulate('submit',{ preventDefault:()=>{} });
    expect(wrapper.state('error').length).toBeGreaterThan(0)
    expect(wrapper).toMatchSnapshot()

})



test('should set descriptiom on input change', ()=>{
    const wrapper = shallow(<ExpenseForm/>)
    //expect(wrapper).toMatchSnapshot()
    wrapper.find('input').at(0).simulate('change', {
        persist:()=>{},
        target:{value:'New description'}
    })
    expect(wrapper.state('description')).toBe('New description')
    //expect(wrapper).toMatchSnapshot()
    
})


test('should set note on input change', ()=>{
    const wrapper = shallow(<ExpenseForm/>)
    wrapper.find('textarea').at(0).simulate('change', {
        persist:()=>{},
        target:{value:'New note'}
    })
    expect(wrapper.state('note')).toBe('New note')
    
})



test('should set amount if valid input', ()=>{
    const wrapper = shallow(<ExpenseForm/>)
    wrapper.find('input').at(1).simulate('change', {
        target:{value:'10.23'}
    })
    expect(wrapper.state('amount')).toBe('10.23')
    
})


test('should not set amount for invalid input', ()=>{
    const wrapper = shallow(<ExpenseForm/>)
    wrapper.find('input').at(1).simulate('change', {
        target:{value:'10.235'}
    })
    expect(wrapper.state('amount')).toBe('')
    
})


test('should call onSubmit prop for valid submission', ()=>{
    const onSubmitSpy = jest.fn()
    const wrapper = shallow(<ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy}/>)
    wrapper.find('form').simulate('submit', {
        preventDefault:()=>{}
    })
    expect(wrapper.state('error').length).toBe(0)
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        description:expenses[0].description,
        amount:expenses[0].amount,
        note:expenses[0].note,
        createdAt:expenses[0].createdAt
    })
    
})


test('should set date on date change', ()=>{
    const wrapper = shallow(<ExpenseForm/>)
    wrapper.find(DatePicker).prop('onChange')(25)
    expect(wrapper.state('createdAt')).toBe(25)
    
})
