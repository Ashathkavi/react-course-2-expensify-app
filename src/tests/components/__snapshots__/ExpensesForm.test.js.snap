// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    typr="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value=""
  />
  <r
    allowSameDay={false}
    className="datepicker"
    closeOnScroll={[Function]}
    customTimeInput={null}
    dateFormat="MM/dd/yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    enableTabLoop={true}
    excludeScrollbar={true}
    focusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next Month"
    nextYearButtonLabel="Next Year"
    onBlur={[Function]}
    onCalendarClose={[Function]}
    onCalendarOpen={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    previousYearButtonLabel="Previous Year"
    readOnly={false}
    renderDayContents={[Function]}
    selected={1970-01-01T00:00:00.000Z}
    shouldCloseOnSelect={true}
    showFourColumnMonthYearPicker={false}
    showFullMonthYearPicker={false}
    showMonthYearPicker={false}
    showPopperArrow={true}
    showPreviousMonths={false}
    showQuarterYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    showTwoColumnMonthYearPicker={false}
    showYearPicker={false}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    withPortal={false}
    yearItemNumber={12}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your expense (optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`should render ExpenseForm with expense data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    typr="text"
    value="Gum"
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value="1.95"
  />
  <r
    allowSameDay={false}
    className="datepicker"
    closeOnScroll={[Function]}
    customTimeInput={null}
    dateFormat="MM/dd/yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    enableTabLoop={true}
    excludeScrollbar={true}
    focusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next Month"
    nextYearButtonLabel="Next Year"
    onBlur={[Function]}
    onCalendarClose={[Function]}
    onCalendarOpen={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    previousYearButtonLabel="Previous Year"
    readOnly={false}
    renderDayContents={[Function]}
    selected={1970-01-01T00:00:00.000Z}
    shouldCloseOnSelect={true}
    showFourColumnMonthYearPicker={false}
    showFullMonthYearPicker={false}
    showMonthYearPicker={false}
    showPopperArrow={true}
    showPreviousMonths={false}
    showQuarterYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    showTwoColumnMonthYearPicker={false}
    showYearPicker={false}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    withPortal={false}
    yearItemNumber={12}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your expense (optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`should render error for invalid form submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    typr="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value=""
  />
  <r
    allowSameDay={false}
    className="datepicker"
    closeOnScroll={[Function]}
    customTimeInput={null}
    dateFormat="MM/dd/yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    enableTabLoop={true}
    excludeScrollbar={true}
    focusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next Month"
    nextYearButtonLabel="Next Year"
    onBlur={[Function]}
    onCalendarClose={[Function]}
    onCalendarOpen={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    previousYearButtonLabel="Previous Year"
    readOnly={false}
    renderDayContents={[Function]}
    selected={1970-01-01T00:00:00.000Z}
    shouldCloseOnSelect={true}
    showFourColumnMonthYearPicker={false}
    showFullMonthYearPicker={false}
    showMonthYearPicker={false}
    showPopperArrow={true}
    showPreviousMonths={false}
    showQuarterYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    showTwoColumnMonthYearPicker={false}
    showYearPicker={false}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    withPortal={false}
    yearItemNumber={12}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your expense (optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`should render error for invalid form submission 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please provide description and amount.
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    typr="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value=""
  />
  <r
    allowSameDay={false}
    className="datepicker"
    closeOnScroll={[Function]}
    customTimeInput={null}
    dateFormat="MM/dd/yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    enableTabLoop={true}
    excludeScrollbar={true}
    focusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next Month"
    nextYearButtonLabel="Next Year"
    onBlur={[Function]}
    onCalendarClose={[Function]}
    onCalendarOpen={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    previousYearButtonLabel="Previous Year"
    readOnly={false}
    renderDayContents={[Function]}
    selected={1970-01-01T00:00:00.000Z}
    shouldCloseOnSelect={true}
    showFourColumnMonthYearPicker={false}
    showFullMonthYearPicker={false}
    showMonthYearPicker={false}
    showPopperArrow={true}
    showPreviousMonths={false}
    showQuarterYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    showTwoColumnMonthYearPicker={false}
    showYearPicker={false}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    withPortal={false}
    yearItemNumber={12}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your expense (optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;
